addons/account/models/account_move_line.py

```diff
@@ -1210,14 +1210,13 @@ def _check_constrains_account_id_journal_id(self):
 
     @api.constrains('account_id', 'tax_ids', 'tax_line_id', 'reconciled')
     def _check_off_balance(self):
-        for line in self:
-            if line.account_id.internal_group == 'off_balance':
-                if any(a.internal_group != line.account_id.internal_group for a in line.move_id.line_ids.account_id):
-                    raise UserError(_('If you want to use "Off-Balance Sheet" accounts, all the accounts of the journal entry must be of this type'))
-                if line.tax_ids or line.tax_line_id:
-                    raise UserError(_('You cannot use taxes on lines with an Off-Balance account'))
-                if line.reconciled:
-                    raise UserError(_('Lines from "Off-Balance Sheet" accounts cannot be reconciled'))
+        for line in self.move_id.line_ids.filtered(lambda l: l.account_id.internal_group == 'off_balance'):
+            if any(a.internal_group != line.account_id.internal_group for a in line.move_id.line_ids.account_id):
+                raise UserError(_('If you want to use "Off-Balance Sheet" accounts, all the accounts of the journal entry must be of this type'))
+            if line.tax_ids or line.tax_line_id:
+                raise UserError(_('You cannot use taxes on lines with an Off-Balance account'))
+            if line.reconciled:
+                raise UserError(_('Lines from "Off-Balance Sheet" accounts cannot be reconciled'))
```

Comment:
```suggestion
        for line in self.move_id.line_ids:
            if line.account_id.internal_group == 'off_balance':
                if any(a.internal_group != line.account_id.internal_group for a in line.move_id.line_ids.account_id):
                    raise UserError(_('If you want to use "Off-Balance Sheet" accounts, all the accounts of the journal entry must be of this type'))
                if line.tax_ids or line.tax_line_id:
                    raise UserError(_('You cannot use taxes on lines with an Off-Balance account'))
                if line.reconciled:
                    raise UserError(_('Lines from "Off-Balance Sheet" accounts cannot be reconciled'))
```
We usually try to minimize the diff (especially in stable versions), so I would recommend something shorter. (1 line)
It's also (very slightly) better in terms of time complexity: O(n) instead of O(n) + O(m) 🤓

Reply:
Thank you for your review.